---
title: "#python の ネストされた辞書で KeyError を起こさないように defaultdict を利用する例 ( #ruby 比較 )"
emoji: "🖥"
type: "idea"
topics: ["Ruby", "Python"]
published: true
---

# example

- defaultdict を何層にもネストして使ってみる
- `d["x"]["y"]["z"]` のように存在しないキーを書いても怒られない
- もしかしたら良い子は真似しちゃダメかもしれない

```py
>>> from collections import defaultdict
>>> d = defaultdict(lambda: defaultdict(lambda: defaultdict(lambda: 0)))
>>> d["a"]["b"]["c"] = 1
>>> d["a"]["b"]["c"]
1
>>> d["x"]["y"]["z"]
0
```

# ruby で書くなら

こんな感じだったかな。こちらも良い子はマネし‥。

```rb
[27] pry(main)> h = Hash.new(Hash.new(Hash.new(0)))
=> {}
[28] pry(main)> h["a"]["b"]["c"] = 1
=> 1
[29] pry(main)> h["a"]["b"]["c"]
=> 1
[30] pry(main)> h["x"]["y"]["z"]
=> 0
```

# ref

[Python defaultdict の使い方 - Qiita](https://qiita.com/xza/items/72a1b07fcf64d1f4bdb7)

# Original by Github issue

https://github.com/YumaInaura/YumaInaura/issues/1284








<!-- Update From Qiita API -->

# チャットメンバー募集


何か質問、悩み事、相談などあればLINEオープンチャットもご利用ください。

https://line.me/ti/g2/eEPltQ6Tzh3pYAZV8JXKZqc7PJ6L0rpm573dcQ





# Twitter


https://twitter.com/YumaInaura


<!-- Update From Qiita API -->


